python/cudf/cudf/core/mixins/mixin_factory.py

```diff
@@ -0,0 +1,218 @@
+            ).parameters.items()
+            if k != "op"
+        ]
+        retfunc.__signature__ = inspect.Signature(retfunc_params)
+
+        setattr(owner, self._name, retfunc)
+
+        if obj is None:
+            return getattr(owner, self._name)
+        else:
+            return getattr(obj, self._name)
+
+
+def _create_delegating_mixin(
+    mixin_name,
+    docstring,
+    category_name,
+    category_operation_name,
+    supported_operations,
+):
+    """Factory for mixins defining collections of delegated operations.
+
+    This function generates mixins based on two common paradigms in cuDF:
+
+    1. libcudf groups many operations into categories using a common API. These
+       APIs usually accept an enum to delineate the specific operation to
+       perform, e.g. binary operations use the `binary_operator` enum when
+       calling the `binary_operation` function. cuDF Python mimics this
+       structure by having operations within a category delegate to a common
+       internal function (e.g. DataFrame.__add__ calls DataFrame._binaryop).
+    2. Many cuDF classes implement similar operations (e.g. `sum`) via
+       delegation to lower-level APIs before reaching a libcudf C++ function
+       call. As a result, many API function calls actually involve multiple
+       delegations to lower-level APIs that can look essentially identical. An
+       example of such a sequence would be DataFrame.sum -> DataFrame._reduce
+       -> Column.sum -> Column._reduce -> libcudf.
+
+    This factory creates mixins for a category of operations implemented by via
+    this delegator pattern. The resulting mixins make it easy to share common
+    functions across various classes while also providing a common entrypoint
+    for implementing the centralized logic for a given category of operations.
+    Its usage is best demonstrated by example below.
+
+    Parameters
+    ----------
+    mixin_name : str
+        The name of the class. This argument should be the same as the object
+        that this function's output is assigned to, e.g.
+        :code:`Baz = _create_delegating_mixin("Baz", ...)`.
+    docstring : str
+        The documentation string for the mixin class.
+    category_name : str
+        The category of operations for which a mixin is being created. This
+        name will be used to define or access the following attributes as shown
+        in the example below:
+            - f'_{category_name}_DOCSTRINGS'
+            - f'_VALID_{category_name}S'
+            - f'_SUPPORTED_{category_name}S'
+    category_operation_name : str
+        The name given to the core function implementing this category of
+        operations.  The corresponding function is the entrypoint for child
+        classes.
+    supported_ops : List[str]
+        The list of valid operations that subclasses of the resulting mixin may
+        request to be implemented.
+
+    Examples
+    --------
+    >>> MyFoo = _create_delegating_mixin(
```

Comment:
I think naming `MyFoo` to `Fooable` to indicate it's an abstraction may add to readability?

Reply:
I think a concrete example (such as reductions) would be even more useful.

Reply:
Sure. Using a concrete example is more informative, and self documents it's application in cudf.

Reply:
I'm going to vote that we continue this discussion but let the PR be merged. The follow-up PRs are already in flight, and we should be able to write much more useful and succinct examples when there are multiple exemplary cases to pull from.

Reply:
Don't mean to belabor the point, but a simpler docstring would be _really_ helpful here. How about using an example like the following?

```python
# The class below:
In [47]: class Person:
    ...:     def _greet(self, op: str):
    ...:         """Say {op}"""
    ...:         print(op)
    ...:
    ...:     def hello(self):
    ...:         self._greet("hello")
    ...:
    ...:     def goodbye(self):
    ...:         self._greet("goodbye")
    ...:
    ...:     def welcome(self):
    ...:         self._greet("welcome")

# can  be rewritten using a delegating mixin as follows:
In [48]: Greeter = _create_delegating_mixin(
    ...:     "Greeter", "", "GREETING", "_greet", {"hello", "goodbye", "welcome"}
    ...: )

In [49]: class Person(Greeter):
    ...:     _VALID_GREETINGS = {"hello", "goodbye", "welcome"}
    ...:
    ...:     def _greet(self, op: str):
    ...:         """Say {op}"""
    ...:         print(op)
```

----

You could even have a `LoudPerson` class that greets in all caps to show inheritance behaviour, for example.


If it's a thumbs down from either of you, happy to resolve this later with another example.

Reply:
Done.